server/src/main/java/org/apache/iotdb/db/rescon/PrimitiveArrayManager.java

```diff
@@ -0,0 +1,329 @@
+import org.apache.iotdb.tsfile.utils.Binary;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manage all primitive data list in memory, including get and release operation.
+ */
+public class PrimitiveArrayManager {
+
+  /**
+   * data type -> ArrayDeque<Array>
+   */
+  private static final Map<TSDataType, ArrayDeque<Object>> bufferedArraysMap = new EnumMap<>(
+      TSDataType.class);
+
+  /**
+   * data type -> current number of buffered arrays
+   */
+  private static final Map<TSDataType, Integer> bufferedArraysNumMap = new EnumMap<>(
+      TSDataType.class);
+
+  /**
+   * data type -> ratio of data type in schema, which could be seen as recommended ratio
+   */
+  private static final Map<TSDataType, Double> bufferedArraysNumRatio = new EnumMap<>(
+      TSDataType.class);
+
+  private static final Logger logger = LoggerFactory.getLogger(PrimitiveArrayManager.class);
+
+  private static final IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+
+  public static final int ARRAY_SIZE = config.getPrimitiveArraySize();
+
+  /**
+   * threshold total size of arrays for all data types
+   */
+  private static final double BUFFERED_ARRAY_SIZE_THRESHOLD =
+      config.getAllocateMemoryForWrite() * config.getBufferedArraysMemoryProportion();
+
+  /**
+   * total size of buffered arrays
+   */
+  private static AtomicInteger bufferedArraysSize = new AtomicInteger();
+
+  /**
+   * total size of out of buffer arrays
+   */
+  private static AtomicInteger outOfBufferArraysSize = new AtomicInteger();
+
+  static {
+    bufferedArraysMap.put(TSDataType.BOOLEAN, new ArrayDeque<>());
+    bufferedArraysMap.put(TSDataType.INT32, new ArrayDeque<>());
+    bufferedArraysMap.put(TSDataType.INT64, new ArrayDeque<>());
+    bufferedArraysMap.put(TSDataType.FLOAT, new ArrayDeque<>());
+    bufferedArraysMap.put(TSDataType.DOUBLE, new ArrayDeque<>());
+    bufferedArraysMap.put(TSDataType.TEXT, new ArrayDeque<>());
+  }
+
+  private PrimitiveArrayManager() {
+  }
+
+  /**
+   * Get primitive data lists according to type
+   *
+   * @param dataType data type
+   * @return an array
+   */
+  public static Object getDataListByType(TSDataType dataType) {
+    // check buffered array num
+    if (bufferedArraysSize.get() + ARRAY_SIZE * dataType.getDataTypeSize()
+        > BUFFERED_ARRAY_SIZE_THRESHOLD) {
+      // return an out of buffer array
+      outOfBufferArraysSize.addAndGet(ARRAY_SIZE * dataType.getDataTypeSize());
+      return getDataList(dataType);
+    }
+
+    synchronized (bufferedArraysMap.get(dataType)) {
+      // return a buffered array
+      bufferedArraysNumMap.put(dataType, bufferedArraysNumMap.getOrDefault(dataType, 0) + 1);
+      bufferedArraysSize.addAndGet(ARRAY_SIZE * dataType.getDataTypeSize());
+      Object dataArray = bufferedArraysMap.get(dataType).poll();
+      if (dataArray != null) {
+        return dataArray;
+      }
+    }
+    return getDataList(dataType);
```

Comment:
We may not get here, so we should return null and add some comments

Reply:
It's my fault .... Could you please help add some comments @HTHou ? This return statement should not reached..

Reply:
> It's my fault .... Could you please help add some comments @HTHou ? This return statement should not reached..
Sure, Let me fix it.

Reply:
I'm confused why should it return null here. If change it, the tests will failed...